File: java/src/0203.java
import static org.assertj.core.api.Assertions.assertThat;
import static structs.util.LinkedList.of;

import java.time.Instant;
import java.time.temporal.ChronoUnit;

import structs.linkedlist.ListNode;

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode removeElements(ListNode head, int val) {
        var fake = new ListNode(0, head);
        var curr = fake;
        while (curr.next != null) {
            if (curr.next.val == val) {
                curr.next = curr.next.next;
            } else {
                curr = curr.next;
            }
        }
        return fake.next;
    }

    public static void main(String[] args) {
        var o = new Solution();

        assertThat(o.removeElements(of(1, 2, 6, 3, 4, 5, 6), 6)).isEqualTo(of(1, 2, 3, 4, 5));
        assertThat(o.removeElements(null, 1)).isNull();
        assertThat(o.removeElements(of(7, 7, 7, 7), 7)).isNull();

        var start = Instant.now();
        for (var i = 0; i < 100; i++) {
            o.removeElements(of(1, 2, 6, 3, 4, 5, 6), 6);
            o.removeElements(null, 1);
            o.removeElements(of(7, 7, 7, 7), 6);
        }
        System.out.println(ChronoUnit.MICROS.between(start, Instant.now()));
    }
}

File: python/src/0203.py
from typing import Optional

from structs.ListNode import ListNode, create


# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeElements(self, head: Optional[ListNode], val: int) -> Optional[ListNode]:
        fake = tail = ListNode(0, head)
        while tail.next:
            if tail.next.val == val:
                tail.next = tail.next.next
            else:
                tail = tail.next
        return fake.next


if __name__ == '__main__':
    import unittest

    CASES = (
        (create(1, 2, 6, 3, 4, 5, 6), 6, create(1, 2, 3, 4, 5)),
        (None, 1, None),
        (create(7, 7, 7, 7), 7, None),
    )

    class Test(unittest.TestCase):
        def test(self):
            removeElements = Solution().removeElements
            for head, val, excepted in CASES:
                actual = removeElements(head, val)
                self.assertEqual(actual, excepted)

    unittest.main()
